File: devicesetup/src/main/java/io/particle/android/sdk/devicesetup/setupsteps/CheckIfDeviceClaimedStep.java
package io.particle.android.sdk.devicesetup.setupsteps;


import java.util.List;

import io.particle.android.sdk.cloud.ParticleCloud;
import io.particle.android.sdk.cloud.ParticleCloudException;
import io.particle.android.sdk.cloud.ParticleDevice;


public class CheckIfDeviceClaimedStep extends SetupStep {

    private final ParticleCloud sparkCloud;
    private final String deviceBeingConfiguredId;
    // FIXME: this seems kind of lame, look into a diff approach to determine this
    private boolean needToClaimDevice;

    public CheckIfDeviceClaimedStep(StepConfig stepConfig, ParticleCloud sparkCloud,
                                    String deviceBeingConfiguredId, boolean needToClaimDevice) {
        super(stepConfig);
        this.sparkCloud = sparkCloud;
        this.deviceBeingConfiguredId = deviceBeingConfiguredId;
        this.needToClaimDevice = needToClaimDevice;
    }

    @Override
    protected void onRunStep() throws SetupStepException {
        List<ParticleDevice> devices;
        try {
            devices = sparkCloud.getDevices();
        } catch (ParticleCloudException e) {
            throw new SetupStepException(e);
        }

        log.d("Got devices back from the cloud...");
        for (ParticleDevice device : devices) {
            if (deviceBeingConfiguredId.equalsIgnoreCase(device.getID())) {
                log.d("Success, device " + device.getID() + " claimed!");
                needToClaimDevice = false;
                return;
            }
        }

        // device not found in the loop
        throw new SetupStepException("Device " + deviceBeingConfiguredId + " still not claimed.");
    }

    @Override
    public boolean isStepFulfilled() {
        return !needToClaimDevice;
    }

}

File: devicesetup/src/main/java/io/particle/android/sdk/devicesetup/setupsteps/CheckIfDeviceClaimedStep.kt
package io.particle.android.sdk.devicesetup.setupsteps


import io.particle.android.sdk.cloud.ParticleCloud
import io.particle.android.sdk.cloud.ParticleDevice
import io.particle.android.sdk.cloud.exceptions.ParticleCloudException


class CheckIfDeviceClaimedStep internal constructor(stepConfig: StepConfig,
                                                    private val sparkCloud: ParticleCloud,
                                                    private val deviceBeingConfiguredId: String) : SetupStep(stepConfig) {
    override val isStepFulfilled: Boolean
        get() = !needToClaimDevice

    private var needToClaimDevice = true

    @Throws(SetupStepException::class)
    override fun onRunStep() {
        val devices: List<ParticleDevice>
        try {
            devices = sparkCloud.devices
        } catch (e: ParticleCloudException) {
            throw SetupStepException(e)
        }

        log.d("Got devices back from the cloud...")
        for (device in devices) {
            if (deviceBeingConfiguredId.equals(device.id, ignoreCase = true)) {
                log.d("Success, device " + device.id + " claimed!")
                needToClaimDevice = false
                return
            }
        }

        // device not found in the loop
        throw SetupStepException("Device $deviceBeingConfiguredId still not claimed.")
    }

}
